### PR TITLE
Functional tests for WP-CLI commands

### LINKED ISSUE #21: Add Behat tests
Perhaps something for the CLI commands

### PR DESCRIPTION
**Pull request by swissspidy**
**Description**

This PR sets up the necessary infrastructure for running Behat tests for the built-in WP-CLI commands.

So far this includes only a test for `wp traduttore info`, but not yet tests for the other commands.

For these, I think we need to figure out how to create a bunch of GlotPress projects with meta data etc. when running the tests. Ideally there would be some proper WP-CLI commands for that, but GlotPress doesn't currently provide any to do such CRUD operations. Perhaps we have to contribute these upstream... 

That would make writing these tests way easier. An alternative was mentioned below:

> You could just import a SQL dump or create a PHP file which is executed once and creates all the things.

Fixes #21.

**How has this been tested?**

By writing tests :-)

**Types of changes**
New functional tests for WP-CLI commands.

**Checklist:**
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by swissspidy** [comment]
The Travis setup [currently fails](https://travis-ci.com/wearerequired/traduttore/jobs/159490430) and I'm not sure why:

```
$ composer dev-build
> rm -rf vendor/autoload.php
> composer dump-autoload
Generating autoload files
> mkdir -p $(pwd)/build
> wp dist-archive "$(pwd)" "$(pwd)/build/traduttore.zip"
Error: 	zip warning: name not matched: traduttore
zip error: Nothing to do! (/home/travis/build/wearerequired/traduttore/build/traduttore.zip)
```

`composer dump-autoload`, `wp dist-archive "$(pwd)" "$(pwd)/build/traduttore.zip"` and `wp dist-archive . "$(pwd)/build/traduttore.zip"` all work locally.

**Comment by swissspidy** [comment]
> In general, we need to figure out how to create a bunch of GlotPress projects with meta data etc. when running the tests. Ideally there would be some proper WP-CLI commands for that, but GlotPress doesn't currently provide any to do such CRUD operations. Perhaps we have to contribute these upstream...

@ocean90 Can you confirm that last part?

**Comment by ocean90** [comment]
All the available CLI commands can be found in https://github.com/GlotPress/GlotPress-WP/tree/develop/gp-includes/cli. Yes, there's nothing for CRUD operations.

Why do you need CLI commands to create the sample data though? You could just import a SQL dump or create a PHP file which is executed once and creates all the things.

**Comment by swissspidy** [comment]
There might be some progress on this in the future, as we're looking into adding such tests for the AMP plugin and evaluating how it can be made easier. https://github.com/ampproject/amp-wp/issues/3079

**Comment by swissspidy** [comment]
See also https://github.com/ampproject/amp-wp/issues/1815#issuecomment-523073046 and https://github.com/ampproject/amp-wp/issues/3076

Using a phar for Behat sounds more straightforward for now. https://github.com/civicrm/composer-downloads-plugin would easily allow for that. Works very well for including things like PHPStan with a phar too.

**Comment by schlessera** [comment]
This has been done now for the AMP plugin here: https://github.com/ampproject/amp-wp/pull/6070
(with code coverage still not 100% done).

**Comment by swissspidy** [comment]
@schlessera Amazing work! 🎉 🚀

**Comment by swissspidy** [comment]
Updating this PR was really straightforward. The tests passed with the first attempt. Now we'd just have to write tests for the more complex commands :-)

Now just have to figure out how to play nicely with the PHPStan WP-CLI stubs. Removing them completely doesn't work.

File: composer.json [hunk 1]
```diff
@@ -77,8 +86,26 @@
...
+    "behat": "behat",
     "format": "vendor/bin/phpcbf --report-summary --report-source .",
     "lint": "vendor/bin/phpcs --report-summary --report-source .",
-    "test": "vendor/bin/phpunit"
+    "phpunit": "vendor/bin/phpunit",
+    "test": [
+      "@analyze",
+      "@behat",
+      "@lint",
+      "@phpunit"
```
Comment (ocean90) [inline composer.json:97]:
```suggestion
      "@behat",
      "@phpunit"
```

I don't think analyze and lint scripts should be part of tests.

File: composer.json [hunk 2]
```diff
@@ -41,12 +41,19 @@
     "php-stubs/wp-cli-stubs": "^2.4",
     "phpunit/phpunit": "^6.5 || ^7.5",
     "szepeviktor/phpstan-wordpress": "^0.7.0",
-    "wearerequired/coding-standards": "^1.5",
+    "wearerequired/coding-standards": "^2.1.0",
```
Comment (ocean90) [inline composer.json:44]:
```suggestion
    "wearerequired/coding-standards": "^1.5",
```

This should not be updated as part of this PR.

**Review by ocean90 — CHANGES_REQUESTED** [review]

**Comment by swissspidy** [comment]
> Now just have to figure out how to play nicely with the PHPStan WP-CLI stubs. Removing them completely doesn't work.

@ocean90 @schlessera any ideas on how to make PHPStan happy here?

**Comment by ocean90** [comment]
Maybe @szepeviktor has an idea? https://github.com/wearerequired/traduttore/pull/124/checks?check_run_id=2522558897#step:9:6

**Comment by szepeviktor** [comment]
> Maybe szepeviktor has an idea?

You make me think!

Please send the usual €1 000 000.
Thank you.

**Comment by szepeviktor** [comment]
Cooking something over at #209 
https://github.com/wearerequired/traduttore/pull/209/checks?check_run_id=2737598412

**Comment by szepeviktor** [comment]
@swissspidy Here is another fix to this branch  #211

**Comment by szepeviktor** [comment]
Where does GHA get its config?
This does not include "Run tests with coverage" but it is apparently executed.
https://github.com/wearerequired/traduttore/actions/runs/903475711/workflow
Please advise.

**Review by szepeviktor — APPROVED** [review]

File: .github/workflows/behat.yml [hunk 3]
```diff
@@ -55,9 +65,18 @@ jobs:
         uses: shivammathur/setup-php@v2
         with:
           php-version: ${{ matrix.php }}
-          coverage: none
+          coverage: ${{ matrix.coverage && 'pcov' || 'none' }}
```
Comment (szepeviktor) [inline .github/workflows/behat.yml:68]: What a solution!

😯

Comment (swissspidy) [reply .github/workflows/behat.yml:68]: Hat tip to WP-CLI!

**Comment by swissspidy** [comment]
So for Behat coverage `maybe-generate-wp-cli-coverage.php` needs to updated for the right version of the `phpunit/php-code-coverage` package.

Also, seems a bit wasteful to run the Behat tests against all these configurations.

File: composer.json [hunk 4]
```diff
@@ -75,10 +81,32 @@
       "url": "https://wpackagist.org"
     }
   ],
+  "minimum-stability": "dev",
+  "prefer-stable": true,
```
Comment (swissspidy) [inline composer.json:85]: Might be able to remove these again

File: composer.json [hunk 5]
```diff
@@ -42,11 +42,17 @@
...
+    "wp-cli/wp-cli-tests": "^3.0.11",
     "wpackagist-plugin/glotpress": "^2.3.1"
   },
   "suggest": {
     "wpackagist-plugin/slack": "Send Slack notifications for various events"
   },
+  "config": {
+    "process-timeout": 7200,
+    "sort-packages": true
+  },
```
Comment (swissspidy) [inline composer.json:55]: Not strictly needed for this PR, but seems useful

File: tests/behat/maybe-generate-wp-cli-coverage.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
+<?php
```
Comment (swissspidy) [inline tests/behat/maybe-generate-wp-cli-coverage.php:1]: This file needs updating

File: .github/workflows/behat.yml [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+      - name: Run tests
+        run: composer run-script behat
+
+      - name: Retrieve list of coverage files
+        if: ${{ matrix.coverage == true }}
+        run: |
+          FILES=$(ls -d -1 "$GITHUB_WORKSPACE/build/logs/clover-behat/"*.* | xargs echo | sed 's/ /,/g')
+          echo $FILES
+          echo "COVERAGE_FILES=$FILES" >> $GITHUB_ENV
```
Comment (swissspidy) [inline .github/workflows/behat.yml:130]: Todo: update step so it fails if no coverage file exists.

Comment (swissspidy) [inline .github/workflows/behat.yml:130]: Might want to use `echo "::set-output name=coverage_files::$FILES"` instead

File: .github/workflows/behat.yml [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+      - name: Upload code coverage report
+        if: ${{ matrix.coverage == true }}
+        uses: codecov/codecov-action@v1.5.0
+        with:
+          files: ${{ env.COVERAGE_FILES }}
+          flags: php,feature
```
Comment (swissspidy) [inline .github/workflows/behat.yml:137]: Just `feature`?

Comment (szepeviktor) [reply .github/workflows/behat.yml:137]: BDD? Gherkin?

File: .github/workflows/behat.yml (hunk 7 above)
Comment (szepeviktor) [reply .github/workflows/behat.yml:130]: Sending #213 ...

File: .distignore [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+.github
+bin
+build
+docs
+tests
+.distignore
+.editorconfig
+.gitattributes
+.gitignore
+.travis.yml
```
Comment (ocean90) [inline .distignore:11]:
```suggestion
```

**Review by szepeviktor — DISMISSED** [review]

**Comment by ocean90** [comment]
Is the code coverage working? Wondering since there's no change on https://app.codecov.io/gh/wearerequired/traduttore/compare/124.

**Comment by szepeviktor** [comment]
> Is the code coverage working?

The upload step in CI says

> Reports have been successfully queued for processing at codecov.io/github/wearerequired/traduttore/commit/0a906c10bb9465558fc77a0cbadac3dfcb8f7706

And Codecov knows about our latest commit
https://github.com/wearerequired/traduttore/pull/124/checks?check_run_id=3027581945

**Comment by swissspidy** [comment]
> > Is the code coverage working?
> 
> The upload step in CI says
> 
> > Reports have been successfully queued for processing at codecov.io/github/wearerequired/traduttore/commit/0a906c10bb9465558fc77a0cbadac3dfcb8f7706
> 
> And Codecov knows about our latest commit
> [#124 (checks)](https://github.com/wearerequired/traduttore/pull/124/checks?check_run_id=3027581945)

Plus, locally the clover files are generated just fine and I can see the code coverage for `InfoCommand` and the like as expected.

Screenshot 2021-07-09 at 12 16 11

There must be something wrong with the paths in there or something.

I'll try to compare the output with the one from PHPUnit.

**Comment by swissspidy** [comment]
So strange 😕 

Setting `pcov.directory` and `pcov.exclude` did definitely help and the coverage files now look more or less the same as when using Xdebug.

The only big difference I found between the coverage file from PHPUnit and the one from Behat is the namespace.

The former has `namespace="Required\Traduttore"` and `<package name="Required\Traduttore">` everywhere, the one from Behat has `namespace="global"`.

**Comment by ocean90** [comment]
fwiw: @schlessera did disable the codecov upload in https://github.com/ampproject/amp-wp/pull/6070/commits/07a844f4f2ee4a7bfc6e7e00af3a883d9c60d342. The comment for the reason isn't clear to me though.

**Comment by swissspidy** [comment]
> fwiw: @schlessera did disable the codecov upload in [ampproject/amp-wp@07a844f](https://github.com/ampproject/amp-wp/commit/07a844f4f2ee4a7bfc6e7e00af3a883d9c60d342). The comment for the reason isn't clear to me though.

Interesting! I don't see why merging wouldn't work though, as we usually merge multiple reports in the web-stories-wp repo and it works just fine there.

**Review by ocean90 — APPROVED** [review]
Let's merge this and see if we can fix the coverage upload in a follow-up PR.

**Pull request merged**